Add Application centric deployment guide

maximiliankolb opened the pull request:
Hi everyone!

This PR provides a guide for **Application Centric Deployment** via the [Foreman ACD](https://github.com/ATIX-AG/foreman_acd) & [Smart Proxy ACD](https://github.com/ATIX-AG/smart_proxy_acd) plugins.

Along with the content, there are also commits to allow building a new "orcharhino" target, including a `.css` file, attributes, and a small extension to the README. I would love some feedback if anything else can be/needs to be adjusted.

Feedback is very welcome. 😄 

Thanks.
Maximilian.

maximiliankolb commented:
For those who want to test ACD: [ACD playbook for ELK](https://github.com/ATIX-AG/acd_playbook_for_elk)

melcorr reviewed (commented):
Hello, @maximiliankolb 

Thank you so much for this contribution! 
I am more than happy to review this for you. 
On what level would you like the review? I can comment, for example, on IBM style standards throughout, or I can simply comment on universal type things that would improve the docs? 
The one thing that I would request, that I think would be very good practice for any new guides that we create here, is if new .adoc files for procedures and concepts are stored in the /common/modules/ directory, rather than the /topics directory. This /topics directory is an old-fashion practice that we will get rid of in all of the docs here eventually also. If you look at the installation guides, for example, there is heavy reuse throughout, and all .adoc modules are in the common/modules folder. If you look at the administration guide, it is a blend, with all new content getting added.

TL;DR you're about to unleash a powerful tool to the community. We have no idea how it will evolve. Let's maximise for reuse, if that's OK with you?

melcorr commented:
One other global comment: I am unsure whether you followed some of the discussions on the recent PRs, but we have moved from ifevals to the more-powerful ifdefs, so please feel free to use them!

maximiliankolb commented:
Thanks @melcorr for having a look at it.

I appreciate any kind of feedback, especially regarding style, conventions, and the nitty gritty of adding a new downstream documentation variant.

I will work on moving the '.adoc' files and fixing the 'ifeval' & report back.

maximiliankolb commented:
Rebased to master; moved content from `topics/` to `common/modules/`; and replaced ~ifeval~ with `ifdef`.

lzap commented on guides/README.md:
```diff
@@ -1,6 +1,6 @@
 # Foreman Guides
 
-This is upstream source code of [Red Hat Satellite 6](https://access.redhat.com/documentation/en-us/red_hat_satellite) documentation.
+This is upstream source code of [Red Hat Satellite 6](https://access.redhat.com/documentation/en-us/red_hat_satellite) documentation as well as parts of the [orcharhino documentation](https://docs.orcharhino.com/or/docs/index.html).
```
Nitpick: I'd expect capital letter for Orcharhino.

maximiliankolb replied:
This comes from our style guide: we never capitalize "orcharhino".

lzap commented on guides/README.md:
```diff
@@ -80,7 +82,7 @@ Please read these guidelines before opening a Pull Request. For more information
 
 The content in this repository is shared between the upstream Foreman community and branded downstream products such as Red Hat Satellite and orcharhino by ATIX.
-Therefore, never write "Foreman" or "Satellite" words directly, but use the following variables:
+Therefore, never write "Foreman", "Satellite", or "orcharhino" words directly, but use the following variables:
```
Oh it looks like this is intentional, okay :-)

lzap commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,43 @@
+You need to install both the `foreman_acd` and `smart_proxy_acd` plugins to use ACD.
+
+This describes the package installation on CentOS 7, Red Hat Enterprise Linux 7, and Oracle Linux 7.
+
```
It would worth explaining that this plugin will work both for Foreman and Katello installations and making an assumption that Foreman (and/or Katello) is installed already - probably referring to the installation guides from here.

maximiliankolb replied:
I've linked both guides and added a warning for Debian and Ubuntu users that ACD is currently untested on Debian based Foreman installations.

lzap commented:
Nice, ignore my comment from the morning. I see you figure it out how to properly add new content. I dropped few random comments, but the final blessing is on @melcorr however some technical notes.

The guide is currently not being built with the proper BUILD flag, you need to edit .github/workflows/deploy.yml. We currently build all combinations of guides and BUILD flags, we want to change this and only build those which are relevant. For now just add two new lines into the GH action.

In order to show the guide on our landing page, you need to edit our static web in web/ and put it there.

maximiliankolb commented:
> The guide is currently not being built with the proper BUILD flag, you need to edit .github/workflows/deploy.yml. We currently build all combinations of guides and BUILD flags, we want to change this and only build those which are relevant. For now just add two new lines into the GH action.
> 
> In order to show the guide on our landing page, you need to edit our static web in web/ and put it there.

Thanks for your input.
I've rebased the PR and addressed both issues; though I am unsure if it's working correctly. Would you mind looking at my changes again?

lzap commented on .github/workflows/deploy.yml:
```diff
@@ -108,6 +110,7 @@ jobs:
           make pdf BUILD=foreman-deb
           make pdf BUILD=katello
           make pdf BUILD=satellite
+          make pdf BUILD=orcharhino
```
Yep this will do it.

lzap commented on guides/common/attributes.adoc:
```diff
@@ -284,3 +284,104 @@ ifdef::katello[]
+:smart-proxy-installation-guide-title: Installing orcharhino Proxy
+:smartproxy-example-com: orcharhino-proxy.example.com
+:smartproxy_port: 8443
+endif::[]
```
My only concern here is now that we have four build types, the list is getting a bit out of control. Only Satellite appears to name thing differently, foreman-el, foreman-deb

lzap replied:
I am working on a patch that should help to clean up the mess, if you woudln't mind dropping your opinion there and then after hopefully it is merged rebasing on top of that.

melcorr commented:
I was on annual leave last week @maximiliankolb and this is on my list for tomorrow morning! Just a head's up.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,50 @@
+[id="{context}_deploying_an_application_instance"]
+= Deploying an Application Instance
+
```
Introduce what the procedure does and why to do it, and any prerequisites.

maximiliankolb replied:
Would it be useful to list configured "Ansible playbook", "application definition", and "application instance" as a requirement?

So far, I've added minimal "one sentence" introductions to each procedure.

melcorr replied:
There is a documentation book called "Every Page is Page One". I am unsure if I read it, because I think that's the entire point of the book. So, if you were to treat the fact that people won't read the docs like a book, but might google and end up on this page, on this section, it would be helpful to have a list of prereqs with links to the relevant sections.

maximiliankolb replied:
I've added a list of requirements to the introduction section.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,50 @@
+xref:{context}_application_instances[Application instances] can be deployed once or multiple times.
+
+.Procedure
+. Navigate to *Applications > App Instances*.
```
```suggestion
. In the {Project} web UI, navigate to *Applications > App Instances*.
```

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,50 @@
+.Procedure
+. Navigate to *Applications > App Instances*.
+. Select the application instance you want to deploy and click the *Deploy* button.
+If the selected application instance has been deployed before, you will be prompted to confirm redeploying existing hosts.
```
```suggestion
If the selected application instance has been deployed before, you are prompted to confirm redeploying to existing hosts.
```

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,50 @@
+== Run Playbook - Customize First
+
+.Procedure
+. Navigate to *Applications > App Instances*.
```
```suggestion
. In the {Project} web UI, navigate to *Applications* > *App Instances*.
```

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,50 @@
+. Navigate to *Applications > App Instances*.
+. Select the application instance you want to deploy and click the *Run Playbook - customize first* button from the drop down menu.
+This opens the *Job Invocation* page with prefilled instructions to deploy the application instance.
+It allows you to customize your application instance deployment and configuration as follows:
```
```suggestion
You can customize your application instance deployment and configuration as follows:
```
IBM Style

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/ref_acd_prerequisites.adoc:
```diff
@@ -0,0 +1,5 @@
+[id="{context}_prerequisites"]
```
I would just include this in the introduction. It's not enough info to stand alone.

melcorr replied:
This feedback still stands?

maximiliankolb replied:
Fixed.

melcorr reviewed (changes requested):
Hi @maximiliankolb 
I got tired making suggestions, so please, wherever you have "Navigate to" as an opening to a procedure, change to `In the {Project} web UI, navigate to` and  please change "via" to using. IBM style asks us not to use Latin terms for translation purposes IIRC. 

In general, I think that the Architecture section is confusing because it contains a definition of components but not really the architeture. I would expect a workflow, or something. 

I think that your terminology section and your architecture sections could be merged into one, as there is overlap here. 

If you strongly disagree with any point I make, I will not force the issue or block the addition of this guide. It's just what I would do.

Thanks to you and ATIX for this contribution.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+[id="{context}_introduction"]
+= Introduction
+
+This guide introduces the ACD plugin and explains how to use it.
```
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+
+This guide introduces the ACD plugin and explains how to use it.
+
+ACD stands for _Application Centric Deployment_ and describes a fundamentally different approach to provisioning and configuring new hosts.
```
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+This guide introduces the ACD plugin and explains how to use it.
+
+ACD stands for _Application Centric Deployment_ and describes a fundamentally different approach to provisioning and configuring new hosts.
+Traditionally, {ProjectName} has pursued a _host centric_ approach.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+
+ACD stands for _Application Centric Deployment_ and describes a fundamentally different approach to provisioning and configuring new hosts.
+Traditionally, {ProjectName} has pursued a _host centric_ approach.
+Now, the ACD approach presents a new idea: deploying hosts in order to run an application that requires a predefined set of services.
```
I was probably sick on that day 😄 

Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+Traditionally, {ProjectName} has pursued a _host centric_ approach.
+Now, the ACD approach presents a new idea: deploying hosts in order to run an application that requires a predefined set of services.
+
+A modern application typically consists of multiple services and therefore requires multiple, different kinds of hosts connected to each other.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+Now, the ACD approach presents a new idea: deploying hosts in order to run an application that requires a predefined set of services.
+
+A modern application typically consists of multiple services and therefore requires multiple, different kinds of hosts connected to each other.
+These hosts have dependencies to each other, e.g. an application server might require a database server.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+You can scale an application instance vertically via compute profile or horizontally by choosing multiple hosts to run the same service to handle increased workloads.
+
+Within ACD, Ansible playbooks are executed on the {ProjectServer}.
+Make sure all managed systems can be reached by Ansible from the {ProjectServer}.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+Deploying and configuring hosts via different {SmartProxyServer}s will be added at a later stage.
+
+[id="{context}_terminology"]
+== Terminology
```
Set to `Key Terms` and adjusted the anchor.

maximiliankolb replied:
Renamed to `Key Terms in Application Centric Deployment` after a discussion on another headline.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+== Terminology
+
+.ACD
+ACD stands for _Application_ _Centric_ _Deployment_ and describes a new approach to provision and configure hosts based on the service they will provide.
```
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+[id="{context}_usage_example"]
+== Usage Example
+
+The following application centric deployment illustrates the deployment of hosts in an application centric approach.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
++
+This example requires four different services: proxy servers, web servers, application servers, and database servers.
+Those hosts are ideally part of a host group.
+. (Optional) Adapt Ansible variables
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
++
+You can either use the default values, change values, or lock the Ansible variables from being edited for an application instance.
+Ansible variables are read from the group variables of the Ansible playbook.
+. (Optional) Overwrite variables from the host group
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+
+[NOTE]
+====
+In case a package needs to be installed via an Ansible playbook and the host is subscribed to {ProjectServer}, make sure the package exists in the content view.
```
Why would you want to capitalize content view?

maximiliankolb replied:
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+[id="{context}_architecture"]
+= Architecture
+
+The ACD plugin is about deploying whole applications instead of single hosts at a time.
```
Removed.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+
+The ACD plugin is about deploying whole applications instead of single hosts at a time.
+
+ACD uses smart proxies and a new remote execution provider called `acd`.
```
True! Fixed and checked for other instances of "new".

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+ACD uses smart proxies and a new remote execution provider called `acd`.
+
+Ansible playbooks are downloaded from {ProjectServer} to the {SmartProxyServer} before being executed.
+You only need to submit the Ansible playbook to the {ProjectServer}, but not an arbitrary number of connected {SmartProxyServer}s.
```
So the idea is that Ansible playbooks are only stored on the ProjectServer and don't have to be sychronizred to any attached SmartProxyServers manually.

This is imporant as in the near future, ACD will also support foreign SmartProxies.

If I understand correctly, other configuration management objects (like Puppet modules etc.) need to by synchronized by hand.

maximiliankolb replied:
Thanks for your input! Applied.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+
+ACD uses smart proxies and a new remote execution provider called `acd`.
+
+Ansible playbooks are downloaded from {ProjectServer} to the {SmartProxyServer} before being executed.
```
Fixed.

FYI: There are a lot of instances: `git grep "the {SmartProxyServer}"`. Should I resolve this in a future PR?

maximiliankolb replied:
Opened a separate PR: https://github.com/theforeman/foreman-documentation/pull/465

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+Ansible playbooks are downloaded from {ProjectServer} to the {SmartProxyServer} before being executed.
+You only need to submit the Ansible playbook to the {ProjectServer}, but not an arbitrary number of connected {SmartProxyServer}s.
+
+This is using a new job provider and new job template.
```
For users, probably not.

Removed.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+The Ansible playbook must contain all necessary roles and tasks to execute the playbook.
+Support for https://docs.ansible.com/ansible/latest/user_guide/collections_using.html[Ansible collections] will be added in a future release.
+
+If your Ansible playbooks are currently using Ansible collections, make sure they are present on every attached {SmartProxyServer} by hand.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+== Application Definitions
+
+Application definitions are blueprints of deployed applications.
+They are required next to an xref:{context}_ansible_playbooks[Ansible playbook] to create xref:{context}_application_instances[application instances].
```
Creating an application instance requires an application definition and an Ansible playbook; so "both side by side" are required.

`next to` > `along with`?

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+An application instance is a preconfigured xref:{context}_application_definitions[application definition] and can be rolled out by end users.
+This automatically deploys and configures hosts based on the application definition and Ansible playbook.
+
+Based on one application definition, end users can configure and deploy multiple application instances.
```
As this is the "architecture" section, we want to make clear that end users only need/can configure an application instance and administrativ users take care of the ansible playbook and application definition.

Therefore, I think it makes sense to use `end users` and `administrative users` which 1:1 correlates to the ACD manager and ACD user roles.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+Navigate to *Applications > App Instances* to manage application instances for application centric deployment.
+
+[id="{context}_ansible_variables_and_parameter_hierarchy"]
+== Ansible Variables and Parameter Hierarchy
```
This section is important to understand how Ansible variables and Foreman parameters are inherited.

I did not find any Foreman documentation on how this works; only [Overriding Ansible Variables](https://www.theforeman.org/plugins/foreman_ansible/2.x/index.html#4.5OverridingAnsibleVariables).

I've added a link to the Ansible docs: [Variable precedence](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable).

Does this make sense?

maximiliankolb replied:
How should I handle this instead?

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+The ACD plugin is used to deploy applications and requires sufficient {ProjectName} permissions.
+Depending on your environment, there might be two different groups of users in regard to _managing_ an application and _deploying_ an application:
+
+* An administrative group of users might be responsible for providing the tools to deploy an application, i.e. an Ansible playbook and application definition.
```
It's more like a "that is" than "for example" in my opinion.
I would argue to leave it as is, as this section specifically addresses user permissions; so `ACD Manager` users have permissions to add, edit, and remove Ansible playbooks and application definitions.

Do you see "for example" because there are more tools necessary to deploy applications?

maximiliankolb replied:
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+[id="{context}_user_roles"]
+== User Roles
+
+The ACD plugin is used to deploy applications and requires sufficient {ProjectName} permissions.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
++
+[source,bash]
+----
+yum install tfm-rubygem-foreman_acd tfm-rubygem-smart_proxy_acd tfm-rubygem-smart_proxy_acd_core
```
Applied (even though we handle thing differently right now).

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
++
+[source,bash]
+----
+foreman-rake db:migrate
```
Applied.

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
++
+[source,bash,subs="+quotes,attributes"]
+----
+{foreman-maintain} service restart
```
Applied.

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
+[source,bash]
+----
+foreman-rake db:migrate
+foreman-rake db:seed
```
Applied.

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
+----
+
+ifdef::foreman-el,foreman-deb,katello,satellite[]
+Refer to the https://theforeman.org/plugins/#2.Installation[Foreman plugin installation] guide for more information.
```
We explicitly use `Refer to X for more information about/on how to do Y` in orcharhino documentation.

I personally dislike `For more information` because it doesn't really feel like a complete sentence.

Nevertheless, fixed (with a dot at the end).

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
+{foreman-maintain} service restart
+----
+
+ifdef::foreman-el,foreman-deb,katello,satellite[]
```
It's not part of the orcharhino documentation.
This MR also adds the `orcharhino` context/build target.

maximiliankolb commented on guides/common/modules/proc_adding_an_ansible_playbook.adoc:
```diff
@@ -0,0 +1,21 @@
+
+.Procedure
+. Navigate to *Applications > Ansible Playbooks*.
+. Click the *New Ansible Playbook* button.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_adding_an_ansible_playbook.adoc:
```diff
@@ -0,0 +1,21 @@
+. In the *Description* field, enter an arbitrary description.
+. In the *Path* field, enter the location of the Ansible playbook on your {ProjectServer}.
++
+We recommend putting the Ansible playbook into `/etc/foreman/plugins/foreman_acd/ansible-playbooks/` to avoid SELinux issues.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,29 @@
+[id="{context}_creating_an_application_definition"]
+= Creating an Application Definition
+
+You can create xref:{context}_application_definitions[application definitions] used for application centric deployment via your {ProjectServer} management UI.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,29 @@
+
+.Procedure
+. Navigate to *Applications > App Definitions*.
+. Click the *New Application Definition* button.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,29 @@
+. Click the *New Application Definition* button.
+. In the *Name* field, enter the name of the new application definition.
+. In the *Description* field, enter an arbitrary description.
+. From the *Ansible Playbook* menu, select the name of an Ansible playbook.
```
Fixed.

maximiliankolb commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,29 @@
+. In the *Name* field, enter the name of the new application definition.
+. In the *Description* field, enter an arbitrary description.
+. From the *Ansible Playbook* menu, select the name of an Ansible playbook.
+. Click the *+* icon to add a new set of hosts.
```
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+
+Within ACD, Ansible playbooks are executed on the {ProjectServer}.
+Make sure all managed systems can be reached by Ansible from the {ProjectServer}.
+Deploying and configuring hosts via different {SmartProxyServer}s will be added at a later stage.
```
Removed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+
+.Ansible Playbooks
+xref:{context}_ansible_playbooks[Ansible playbooks] are the core of the application definition.
+They describe the relation between different services to provide a whole application with one click of a button.
```
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+Deploying and configuring hosts via different {SmartProxyServer}s will be added at a later stage.
+
+[id="{context}_terminology"]
+== Terminology
```
I've settled on `host`.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+xref:{context}_application_definitions[Application definitions] are blueprints of how to deploy an application and its services.
+
+.Application Instances
+xref:{context}_application_instances[Application instances] are configured application definitions.
```
Application instances are always created by the end users themselves.
It's basically saying "how many" hosts for which service and setting unique host names. Configuration using parameters or Ansible variables is optional.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+For example, a system deploying a wiki platform for 20 users might use the `1-small` compute profile, whereas a system with hundreds of expected users might choose the `3-large` compute profile.
+
+.Service
+A service is a type of host necessary to run an application.
```
`Service` is ACD terminology. Each application definition has an optional service count. > you can define how many hosts of each type a user can deploy. For example, you can allow users to deploy something between two and five hosts running a database.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+For example, the _LAMP_ stack consists of three services: an https://httpd.apache.org/[Apache] webserver, a https://www.mysql.com/[MySQL] database, and https://www.php.net/[PHP] scripting language all running on Linux servers.
+
+.Host Groups
+Host groups are predefined sets of provisioning and configuration settings for a new host.
```
Yes, it's the same.

I've added a conditional link.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+
+.Host Groups
+Host groups are predefined sets of provisioning and configuration settings for a new host.
+They allow you to quickly deploy hosts with identical settings with the click of a button.
```
Applied.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+Host groups are predefined sets of provisioning and configuration settings for a new host.
+They allow you to quickly deploy hosts with identical settings with the click of a button.
+
+.Host Centric vs. Application Centric Deployment
```
Moved.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+|===
+
+[id="{context}_usage_example"]
+== Usage Example
```
Renamed and switched `Key Terms` and `ACD Workflow Example`.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+[CAUTION]
+====
+The Ansible playbook must contain all necessary roles and tasks to execute the playbook.
+Support for https://docs.ansible.com/ansible/latest/user_guide/collections_using.html[Ansible collections] will be added in a future release.
```
Removed the promise.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+
+[CAUTION]
+====
+The Ansible playbook must contain all necessary roles and tasks to execute the playbook.
```
This is something that has to be proven helpful for new users of ACD.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+Application definitions are typically handled by administrative {ProjectName} users.
+They serve as a _blueprint_ providing easy self service for end users.
+
+Navigate to *Applications > App Definitions* to manage application definitions for application centric deployment.
```
Removed `typically` to clarify between user roles.

Application definition cannot be added, edited, or removed by ACD users (users having this Foreman role).

I don't know how users use the plugin in combination with Foreman user roles, but it provides two: ACD Manager and ACD User (only allowed to manage application instances).

Renamed `manage`.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+Application definitions are blueprints of deployed applications.
+They are required next to an xref:{context}_ansible_playbooks[Ansible playbook] to create xref:{context}_application_instances[application instances].
+
+Application definitions are typically handled by administrative {ProjectName} users.
```
No change.

I would argue this is fine as is; it's an overview of what ACD is and how it's built. User permissions (i.e. necessary user roles) become clear in the `User Roles` section on the same page.

melcorr replied:
The main problem with this sentence is how passive it is. For me, documentation should be written in an active voice where at all possible. It speaks to no one.

maximiliankolb replied:
Reworded: `Users with administrative access to {ProjectName} handle application definitions and specify the allowed number of hosts for each service.`.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+[id="{context}_application_definitions"]
+== Application Definitions
+
+Application definitions are blueprints of deployed applications.
```
Reworded the app def section.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+[id="{context}_application_instances"]
+== Application Instances
+
+An application instance is a preconfigured xref:{context}_application_definitions[application definition] and can be rolled out by end users.
```
Reworded.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+
+Based on one application definition, end users can configure and deploy multiple application instances.
+
+Navigate to *Applications > App Instances* to manage application instances for application centric deployment.
```
Reworded to "add, deploy, and remove".

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+Depending on your environment, there might be two different groups of users in regard to _managing_ an application and _deploying_ an application:
+
+* An administrative group of users might be responsible for providing the tools to deploy an application, i.e. an Ansible playbook and application definition.
+* An end user group might be using {ProjectName} to deploy applications, i.e. creating and deploying application instances based on provided application definitions within their organization and location context.
```
I would not change this because it is part of the architecture section.

I've removed `provided`.

maximiliankolb commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,117 @@
+
+* An administrative group of users might be responsible for providing the tools to deploy an application, i.e. an Ansible playbook and application definition.
+* An end user group might be using {ProjectName} to deploy applications, i.e. creating and deploying application instances based on provided application definitions within their organization and location context.
+Many application instances can be created based on the same application definition.
```
Not removed as this shows that end users can create and deploy multiple applications of the same type.

maximiliankolb commented on guides/common/modules/proc_acd_installation.adoc:
```diff
@@ -0,0 +1,50 @@
+[id="{context}_installation"]
+= Installation
```
No other headline references `ACD`, so I think we should keep this as is. How do you feel about it/why did you propose this change?

melcorr replied:
We use descriptive headings beginning with a gerund throughout all our guides. I understand what you're saying but it is out of sync with all the other guides, including the installation guides.

maximiliankolb replied:
Fixed.

maximiliankolb commented on guides/common/modules/proc_adding_an_ansible_playbook.adoc:
```diff
@@ -0,0 +1,21 @@
+You can add xref:{context}_ansible_playbooks[Ansible playbooks] used for application centric deployment via your {ProjectServer} management UI.
+
+.Procedure
+. Navigate to *Applications > Ansible Playbooks*.
```
Fixed.

maximiliankolb commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,29 @@
+You can create xref:{context}_application_definitions[application definitions] used for application centric deployment via your {ProjectServer} management UI.
+
+.Procedure
+. Navigate to *Applications > App Definitions*.
```
Is it really necessary to exclude `>`? If so, I can change this in the whole MR later on.

maximiliankolb commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,29 @@
+. In the *Description* field, enter an arbitrary description.
+. From the *Ansible Playbook* menu, select the name of an Ansible playbook.
+. Click the *+* icon to add a new set of hosts.
+** In the *Name* field, enter the name of the new set of hosts.
+** In the *Description* field, enter an arbitrary description.
+** From the *Hostgroup* menu, select an existing host group.
++
+You cannot create a service for an application definition without any existing host groups.
+** From the *Ansible Group* menu, select existing Ansible group variables from the selected Ansible playbook.
+** In the *min count* field, enter the minimum number of hosts providing this service.
+If it is unset, users can choose to not deploy a host providing this service at all.
+** In the *max count* field, enter the maximum number of hosts providing this service.
+If it is unset, users can choose to deploy an unlimited number of hosts providing this service.
+** In the *Actions* column,
+*** click the *edit* icon to edit an entry;
+*** click the *settings* icon to add or edit existing {ProjectName} parameters.
+Every hierarchy level of {ProjectName} parameters can be overwritten.
+*** click the *A* character to edit, add, lock, or delete Ansible group variables;
+*** click the *delete* icon to delete an entry.
+. Click *Submit* to save your application definition.
```
I'm afraid the nested procedure is necessary as most (all?) applications consist of more than one service. This means, users need to run the "level 2 steps" more than once. Having everything is one "linear" list is not sufficient. I've reworded it to `add new sets of hosts` to make this clearer.

I've changed this to ordered list items and created a `Prerequisites` section.

Do you have strong objections to this?

melcorr replied:
I will recover from this ;-) 
If you want, you could use substeps rather than bullets :)

maximiliankolb replied:
I don't know what you mean by substeps. Is there an example in the Foreman docs?

melcorr replied:
You changed here to substeps: https://github.com/theforeman/foreman-documentation/pull/453/files#r604223391

maximiliankolb replied:
Using substeps.

maximiliankolb commented:
Rebased and resolved most comments.

melcorr commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,113 @@
+. Optional: Overwrite variables from the host group
++
+For example, change the compute profile or set a different lifecycle environment.
+. Deploy an application instance
++
+** You may simply deploy the application instance;
+** run the Ansible playbook as is;
+** or customize the Ansible playbook before running it.
```
What's happening here? It reads like a haiku? 
Bullets are for unordered lists. This is a sentence?

maximiliankolb replied:
Best I could do in five minutes:

```
I val-ue this much
con-struc-tive cri-ti-sis-m
re-word-ed glad-ly
```

Reworded.

melcorr commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,113 @@
+
+[NOTE]
+====
+If you want to install a package using an Ansible playbook and the host is subscribed to {ProjectServer}, ensure that the package exists in the content view.
```
Content View is capitalised in Foreman documentation.

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,116 @@
+
+[source,bash]
+----
+ansible-galaxy collection install namespace.collection
```
```suggestion
# ansible-galaxy collection install _namespace.collection_
```

In Foreman docs, we wrap all variables in italics.

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/con_acd_architecture.adoc:
```diff
@@ -0,0 +1,116 @@
+The Ansible playbook must contain all necessary roles and tasks to execute the playbook.
+
+If your Ansible playbooks are currently using Ansible collections, ensure that these Ansible collections are added manually to every {SmartProxyServer}.
+You can install Ansible collections by hand by running the following command on your {ProjectServer} and {SmartProxyServer}:
```
```suggestion
 To install Ansible collections manually, enter the following command on your {ProjectServer} and {SmartProxyServer}:
```
A few things here. 
This "by hand" sounds very strange in English. We generally use the term "manually". 
We always use "enter" for a command and "run" for a script.

maximiliankolb replied:
Thanks, applied!

melcorr commented on guides/common/modules/proc_removing_an_application_instance.adoc:
```diff
@@ -0,0 +1,12 @@
+. In the {ProjectWebUI}, navigate to *Application > App Instances*.
+. In the list of application instances, open the drop down menu and select *Delete* on the application instance you want to remove.
++
+This prompts a list of hosts that have been deployed via the selected application instance.
+You can choose to also delete hosts when deleting the application instance.
+Otherwise, those managed hosts remain.
```
In general, we do not describe anything but the user actions, unless the user has to interact with, for example, the output. We trust that the user will see the list of hosts. I would remove this as it is out of sync with everything else.

maximiliankolb replied:
The user has to interact with it. There's a prompt to select individual hosts to be deleted. Else, hosts stay connected to Foreman.

No change.

melcorr commented on guides/common/modules/proc_creating_an_application_instance.adoc:
```diff
@@ -0,0 +1,26 @@
+. In the {ProjectWebUI}, navigate to *Application > App Instances* and click the *New Application Instance* button.
+. In the *Name* field, enter the name of the new application instance.
+. In the *Description* field, enter an arbitrary description.
+. From the *Application Definition* menu, select an existing application definition.
+. In the table, connect the hosts and necessary services:
+** In the *Hostname* field, enter a unique name for the host.
```
Can you add substeps rather than bullets.

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+====
+
+[id="{context}_deploying_an_application_instance_run_playbook"]
+== Run Playbook
```
```suggestion
== Running the Ansible Playbook
```

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+This will automatically start configuring the specified hosts according to the Ansible playbook.
+
+[id="{context}_deploying_an_application_instance_run_playbook_customize_first"]
+== Run Playbook - Customize First
```
```suggestion
== Customizing an Ansible Playbook to Reconfigure Hosts
```

melcorr replied:
Ideally we are as clear as possible in our titles. Feel free to rework this, but please do more than use the button name because it doesn't quite capture what is going on.

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+.Procedure
+. In the {ProjectWebUI}, navigate to *Applications > App Instances*.
+. Select the application instance you want to deploy and click the *Run Playbook - customize first* button from the drop down menu.
+This opens the *Job Invocation* page with prefilled instructions to deploy the application instance.
```
```suggestion

```
Trust the user and the UX team here. This wastes their time reading. It's anti-minimalist.

maximiliankolb replied:
Removed.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+This opens the *Job Invocation* page with prefilled instructions to deploy the application instance.
+You can customize your application instance deployment and configuration as follows:
++
+** You can make the Ansible playbook output more verbose.
+** You can add more Ansible variables before executing the Ansible playbook.
+** You can set the amount of concurrently running Ansible tasks.
+** You can select specific tags and hereby limit the tasks of the Ansible playbook that will be executed.
```
```suggestion
* You can make the Ansible playbook output more verbose.
* You can add more Ansible variables before executing the Ansible playbook.
* You can set the amount of concurrently running Ansible tasks.
* You can select specific tags and hereby limit the tasks of the Ansible playbook that will be executed.
```

maximiliankolb replied:
This does not result in a visual change.
Is there an explicit convention to handle is this way?

Changed.

melcorr commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,113 @@
+There might be different configurations based on a testing or production environment or its expected usage.
+For example, deploying a wiki platform for 20 users might use the `1-small` compute profile, whereas hundreds of expected users might require the `3-large` compute profile.
+
+.Service
```
I know you explained this, but I can't help but think this is going to cost ATIX in customer support. 
Any other name would help distinguish things and make thing clearer for the user and for your customers. 
I've seen poor naming conventions like this cause confusion for customers in different projects throughout the years. 
I just don't understand why this is the only name there could be for " a type of host necessary to run an application". 
I won't block it going in docs, and this is the last time I'll say anything but we can definitely do better here.

quba42 replied:
Off the top of my head, I can't think of any other word for "a group of hosts of a type that collectively provide a single service", can you provide some ideas?

melcorr replied:
I talked with Tomer yesterday and he suggested "Service Provider" and I would further suggest "Service-providing host" or something like that.

maximiliankolb replied:
We've discussed this internally once more.

We want to keep the term `service`.

melcorr replied:
> We've discussed this internally once more.

In the spirit of our open source project, I would like to follow the logic of your discussions, because right now, I don't understand the _why_ behind this decision. 
Can you help me?

maximiliankolb replied:
I can change line 104 to `A service is a component of an application running on one to many hosts of the same type.`

When creating an application definition, a service is defined from a host group and Ansible group.
When creating an application instance, you must assign hosts to services. According to the service counts of the application definition, one service is provided by zero to many hosts.

melcorr replied:
Hi @maximiliankolb that sounds like an improvement. 
I don't want to hold up this any longer. We can see if this issue comes up anywhere else in the community, and if so, we can revisit. 
Thanks for your patience.

melcorr commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,30 @@
+If it is unset, users can choose to not deploy a host providing this service at all.
+.. In the *max count* field, enter the maximum number of hosts providing this service.
+If it is unset, users can choose to deploy an unlimited number of hosts providing this service.
+.. In the *Actions* column,
```
```suggestion
.. In the *Actions* column, complete the following steps: 
```
use complete sentence.

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,30 @@
+.. In the *max count* field, enter the maximum number of hosts providing this service.
+If it is unset, users can choose to deploy an unlimited number of hosts providing this service.
+.. In the *Actions* column,
+... click the *edit* icon to edit an entry;
```
```suggestion
... Click the *edit* icon to edit an entry;
```

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,30 @@
+If it is unset, users can choose to deploy an unlimited number of hosts providing this service.
+.. In the *Actions* column,
+... click the *edit* icon to edit an entry;
+... click the *settings* icon to add or edit existing {ProjectName} parameters.
```
```suggestion
... Click the *settings* icon to add or edit existing {ProjectName} parameters.
```

melcorr replied:
Are all of these icons really lower case in the web UI?

maximiliankolb replied:
It doesn't say `settings` anywhere, it's an icon. Hovering it shows `change paramters`.

melcorr commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,30 @@
+... click the *edit* icon to edit an entry;
+... click the *settings* icon to add or edit existing {ProjectName} parameters.
+Every hierarchy level of {ProjectName} parameters can be overwritten.
+... click the *A* character to edit, add, lock, or delete Ansible group variables;
```
```suggestion
... Click the *A* character to edit, add, lock, or delete Ansible group variables;
```

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,30 @@
+... click the *settings* icon to add or edit existing {ProjectName} parameters.
+Every hierarchy level of {ProjectName} parameters can be overwritten.
+... click the *A* character to edit, add, lock, or delete Ansible group variables;
+... click the *delete* icon to delete an entry.
```
```suggestion
... Click the *delete* icon to delete an entry.
```

maximiliankolb replied:
Fixed.

melcorr commented on guides/common/modules/proc_creating_an_application_instance.adoc:
```diff
@@ -0,0 +1,26 @@
+. From the *Application Definition* menu, select an existing application definition.
+. In the table, connect the hosts and necessary services:
+** In the *Hostname* field, enter a unique name for the host.
+You may choose from all lowercase characters, digits, and hyphens.
```
```suggestion
You can enter lowercase characters, digits, and hyphens.
```

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_creating_an_application_instance.adoc:
```diff
@@ -0,0 +1,26 @@
+. In the table, connect the hosts and necessary services:
+** In the *Hostname* field, enter a unique name for the host.
+You may choose from all lowercase characters, digits, and hyphens.
+Hosts may not start with a hyphen.
```
```suggestion
Hosts cannot start with a hyphen.
```
From IBM style guide: 

"“May” is overused in technical information and should in most instances be changed to “can” or “might.”"

maximiliankolb replied:
Thanks, applied!

melcorr commented on guides/common/modules/proc_creating_an_application_instance.adoc:
```diff
@@ -0,0 +1,26 @@
+** In the *Hostname* field, enter a unique name for the host.
+You may choose from all lowercase characters, digits, and hyphens.
+Hosts may not start with a hyphen.
+** In the *Description* field, enter an arbitrary description.
```
```suggestion
** In the *Description* field, enter a description.
```

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+.Procedure
+. In the {ProjectWebUI}, navigate to *Applications > App Instances*.
+. Select the application instance you want to deploy and click the *Deploy* button.
+If the selected application instance has been deployed before, you are prompted to confirm redeploying to existing hosts.
```
```suggestion
If the selected application instance has been deployed before, confirm that you want to redeploy to existing hosts.
```

maximiliankolb replied:
Applied.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+. In the {ProjectWebUI}, navigate to *Applications > App Instances*.
+. Select the application instance you want to deploy and click the *Deploy* button.
+If the selected application instance has been deployed before, you are prompted to confirm redeploying to existing hosts.
+Confirming this automatically starts provisioning the specified hosts and configuring them according to the Ansible playbook.
```
```suggestion

```
Unnecessary

maximiliankolb replied:
Removed.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+.Procedure
+. In the {ProjectWebUI}, navigate to *Applications > App Instances*.
+. Select the application instance you want to deploy and click the *Run Playbook* button from the drop down menu.
+This will automatically start configuring the specified hosts according to the Ansible playbook.
```
```suggestion

```

maximiliankolb replied:
Removed.

melcorr commented on guides/common/modules/proc_deploying_an_application_instance.adoc:
```diff
@@ -0,0 +1,54 @@
+
+[TIP]
+====
+You may also choose to deploy the application instance at a certain point of time in the future or recurringly.
```
```suggestion
You can also choose to deploy the application instance at a certain point of time in the future or recurringly.
```

melcorr reviewed (changes requested):
Hi @maximiliankolb 

Something happened in master.adoc that has made the navbar for the guide disappear. Please take a look and fix. 

I don't want to block this any further. I've included a few references here and a few further comments, but I will leave it up to you. 

This is what IBM Style says about "via"

via prep
Use this term, which means “by way of,” to refer to routing data, such as in “Messages are routed via
SMTP.” In other situations, try to replace “via” with one of the following synonyms: “across,” along,” “by,”
“from,” “on,” or “through.”

It's also IBM standard for references to say "For more information about x, see y" 
It's also IBM standard to not use ie or eg, but replace with "for example" or reread. 
In Red Hat, we use "ensure" to replace "make sure" everywhere, for minimalist purposes. 

We also don't describe for the users what appears on the screen, only the actions they take. I have pointed to this in your doc. This is to do with minimalist documentation principles. There is an internal group in RH that focuses on this. I studied this in university also. There are resources on the internet about this, but I do not have a resource at hand. 

If you could check for this globally also, as per IBM style:

“May” is overused in technical information and should in most instances be changed to “can” or “might.”

Again, Maximilian, thank you for your work. I will not block this any further. I am just trying to manage the inconsistencies so that we have a standard we expect from all doc contributions. If there are inconsistencies, one can always point to that as precedent, which is unhelpful.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,112 @@
+[id="{context}_introduction"]
+= Introduction
```
I've added this to my todo list, although I don't think we will prioritize this in the near future.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -1,48 +1,18 @@
 = Introduction
 
-This guide introduces the ACD plugin and explains how to use it.
+_Application Centric Deployment_ (ACD) and describes a fundamentally different approach to provisioning and configuring hosts in {Project}.
```
Fixed.

maximiliankolb commented on guides/common/modules/con_acd_introduction.adoc:
```diff
@@ -0,0 +1,113 @@
+== Key Terms
+
+.ACD
+ACD stands for _Application_ _Centric_ _Deployment_ and describes a new approach to provision and configure hosts based on the service they provide.
```
Removed `and`, `e.g.`, and `i.e.`.

melcorr commented on guides/common/modules/proc_creating_an_application_definition.adoc:
```diff
@@ -0,0 +1,30 @@
+. In the *Description* field, enter an arbitrary description.
+. From the *Ansible Playbook* list, select the name of an Ansible playbook.
+. Click the plus icon to add new sets of hosts.
+.. In the *Name* field, enter the name of the new set of hosts.
```
@maximiliankolb here you have substeps rather than bullets and I think it looks good.

maximiliankolb commented:
Rebased to master & [fixed some more issues](https://github.com/theforeman/foreman-documentation/pull/453/commits/3ffa7b506d3bc9010825553b33ff2da277937910).

I am not aware of any issue with the narbar; to me, the head of 'master.adoc' looks very similar to other guides.

melcorr commented:
hi @maximiliankolb 
> I am not aware of any issue with the narbar; to me, the head of 'master.adoc' looks very similar to other guides.

When you build locally, do you see a menu bar?

maximiliankolb commented:
> hi @maximiliankolb
> 
> > I am not aware of any issue with the narbar; to me, the head of 'master.adoc' looks very similar to other guides.
> 
> When you build locally, do you see a menu bar?

Fixed!

maximiliankolb commented:
I have no idea why the link check is failing. I feel like it's connected to the build target `orcharhino` and invalid attribute substitutions. Can someone please help me debug this?

lzap commented:
Hey, our link checker ignores all URLs which are `example.com`, I think you need do change `{Keycloak-short}` URLs to `{Keycloak-short}.example.com`. Not sure why this happens in this PR, perhaps we merged this by accident into master and you rebased on top of it?

melcorr commented:
Are you able to build guides locally since the last update? I can't.

maximiliankolb commented:
> Are you able to build guides locally since the last update? I can't.

I can build in my Gitlab Pipeline with `cd guides/ && make html` just fine. Debugging the failed HTML link checker task was quite difficult but I think it's working now.

melcorr reviewed (commented):
Hi @maximiliankolb 
Everything is building fine now with your last changes, and on review, everything looks good. 
I have talked to a few people regarding the Service issue, and I really think that this must be changed. We can bring this up on the Foreman community Discourse if you would like further input from others. I made a suggestion in the comments here. 
Other than that I think we are good to go. 
I apologise for the inconvenience, but from experience, it is important to get these things correct early on. I appreciate your (plural) openness to suggestions about this.

maximiliankolb commented:
Thanks @melcorr for your input & patience!

I've rebased it once more. From my side it's good to go.

lzap commented:
I just checked `master` and it is green on GH Actions and locally it builds fine. Not sure why you see those errors but we should definitely use `example.com` domains for all example URLs.

lzap commented:
Hello,

we have merged a big cleanup in `attributes.adoc` file, it is now split into several files:

* [attributes.adoc](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes.adoc): version definitions and includes for other attribute files.
* [attributes-base.adoc](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc): base attributes common for all builds.
* [attributes-foreman-el.adoc](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-foreman-el.adoc): base overrides for foreman-el build.
* [attributes-foreman-deb.adoc](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-foreman-deb.adoc): base overrides for foreman-deb build.
* [attributes-katello.adoc](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-katello.adoc): base overrides for katello build.
* [attributes-satellite.adoc](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-satellite.adoc): base overrides for satellite build.

Please resolve the conflicts and update the attributes accordingly, sorry for this but we believe this will ultimately lead to cleaner attributes. I wish you nice Easter!

maximiliankolb commented:
Rebased to master.

maximiliankolb commented:
What can I do to move this forward?

melcorr reviewed (approved):
Let's merge. Thank you so much for your patience with me @maximiliankolb
